File: bindings/rust/tests/stream_sticky.rs
//! A decrypt session fed a tampered wire fails with a sticky error.

use itb::{OptsBuilder, Pipeline};

#[test]
fn tampered_wire_sticky_failure() {
    let opts = OptsBuilder::new();
    let sender = Pipeline::init("streaming-aead-triple-mac-v1", &opts).unwrap();
    let receiver =
        Pipeline::open("streaming-aead-triple-mac-v1", sender.blob(), &opts, None).unwrap();

    let plain: Vec<u8> = (0..65_536).map(|i| (i % 227) as u8).collect();
    let mut wire = sender.encrypt_stream_one_shot(&plain).unwrap();

    // Flip one bit inside the wire body.
    let idx = wire.len() * 3 / 4;
    wire[idx] ^= 0x01;

    let mut sess = receiver.decrypt_stream().unwrap();
    // The failure may surface on write (chain already failed) or on a
    // later read — either way a read must eventually report it.
    let _ = sess.write(&wire);
    let _ = sess.end();

    let mut buf = [0u8; 4096];
    let mut first_err = None;
    loop {
        match sess.read(&mut buf) {
            Ok((_, true)) => break,
            Ok((_, false)) => continue,
            Err(e) => {
                first_err = Some(e);
                break;
            }
        }
    }
    let first = first_err.expect("tampered wire must fail authentication / decryption");
    let first_status = first.status().expect("error carries a status code");

    // Sticky: a subsequent read reports the same status.
    let again = sess.read(&mut buf).unwrap_err();
    assert_eq!(again.status(), Some(first_status));
}

File: bindings/zig/tests/stream_sticky.zig
//! A decrypt session fed a tampered wire fails with a sticky
//! `error.MacFailure`.
//!
//! A single bit flip can land in the container's CSPRNG residue —
//! over-sized container area that carries no payload — where the
//! decrypt legitimately completes clean. The test therefore probes
//! successive flip positions, each against a fresh session on a
//! fresh copy of the wire, until one lands in authenticated content;
//! the observed failure must be MAC failure and must be sticky. The
//! probe is black-box — no wire-layout knowledge is used.

const std = @import("std");
const itb = @import("itb");

/// Feeds one tampered wire copy through a fresh decrypt session.
/// Returns false when the session finishes clean (flip landed in
/// unauthenticated residue), true when it failed with the expected
/// sticky MAC failure; any other failure shape is a test error.
fn probeOnce(
    gpa: std.mem.Allocator,
    receiver: *itb.Pipeline,
    wire: []const u8,
    flip_pos: usize,
) !bool {
    const tampered = try gpa.dupe(u8, wire);
    defer gpa.free(tampered);
    tampered[flip_pos] ^= 0x01;

    var sess = try receiver.decryptStream();
    defer sess.deinit();

    // The failure may surface on write (chain already failed) or on
    // a later read — either way a read must eventually report it.
    sess.write(tampered) catch {};
    sess.end() catch {};

    var buf: [4096]u8 = undefined;
    while (true) {
        const r = sess.read(&buf) catch |first| {
            try std.testing.expectEqual(error.MacFailure, first);
            // Sticky: a subsequent read reports the same error.
            try std.testing.expectError(error.MacFailure, sess.read(&buf));
            return true;
        };
        if (r.finished) return false;
    }
}

test "tampered wire fails with sticky MacFailure" {
    const gpa = std.testing.allocator;

    var sender = try itb.Pipeline.init(gpa, "streaming-aead-triple-mac-v1", null);
    defer sender.deinit();
    var receiver = try itb.Pipeline.open(
        gpa,
        "streaming-aead-triple-mac-v1",
        sender.blob(),
        null,
        null,
    );
    defer receiver.deinit();

    const size: usize = 65536;
    const plain = try gpa.alloc(u8, size);
    defer gpa.free(plain);
    for (plain, 0..) |*b, i| b.* = @truncate(i % 227);

    const wire = try sender.encryptStreamPump(plain);
    defer gpa.free(wire);
    try std.testing.expect(wire.len > 0);

    var seen_failure = false;
    var attempt: usize = 0;
    while (attempt < 32 and !seen_failure) : (attempt += 1) {
        const flip_pos = (wire.len * 3 / 4 + attempt * 1031) % wire.len;
        seen_failure = try probeOnce(gpa, &receiver, wire, flip_pos);
    }
    try std.testing.expect(seen_failure);
}
